### PR TITLE
Snap 2772

### PR DESCRIPTION
**Pull request by smahajan05**
## Changes proposed in this pull request

- Added code changes for undeploy package and jar from cluster.
- Undeploy Jar/Package now removes the jars from snappy executors and unloads the classes from lead urlclassloader and snappy executor classloader

## Patch testing

Manually tested and will be adding tests in test suite
Precheckin is running

## ReleaseNotes.txt changes

None

## Other PRs 
https://github.com/SnappyDataInc/spark/pull/162

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/ddl.scala [hunk 1]
```diff
@@ -628,6 +626,40 @@ case class ListPackageJarsCommand(isJar: Boolean) extends RunnableCommand {
...
+      val cmndsSet = ToolsCallbackInit.toolsCallback.getGlobalCmndsSet
+      cmndsSet.forEach(new Consumer[Entry[String, String]] {
+        override def accept(t: Entry[String, String]): Unit = {
+          val alias1 = t.getKey
+          if(alias == alias1) {
+            value = t.getValue
+          }
+        }
+      })
+      val indexOf = value.indexOf("|")
```
Comment (ashetkar) [inline core/src/main/scala/org/apache/spark/sql/execution/ddl.scala]: Please see if you can re-use the code from `SnappyContext.initGlobalSnappyContext()` and avoid code duplication.

File: cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala [hunk 2]
```diff
@@ -178,6 +176,30 @@ class SnappyExecutor(
     }
   }
 
+  def removeJarsFromExecutorLoader(jars: Array[String]): Unit = {
+    synchronized {
+      jars.foreach(name => {
+        val localName = name.split("/").last
+        var jarFile = new File(SparkFiles.getRootDirectory(), localName)
+        if (jarFile.exists()) {
+          jarFile.delete()
```
Comment (ashetkar) [inline cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala]: For DEPLOY JAR, this deletes the jar file copied to executor/server's work directory.
What is the behavior for DEPLOY PACKAGE?

File: cluster/src/main/scala/io/snappydata/ToolsCallbackImpl.scala [hunk 3]
```diff
@@ -209,4 +215,19 @@ object ToolsCallbackImpl extends ToolsCallback with Logging {
...
+    val lead = ServiceManager.getLeadInstance.asInstanceOf[LeadImpl]
+    val allURLs = lead.urlclassloader.getURLs
+    val updatedURLs = allURLs.toBuffer
+    uris.foreach(uri => {
+      val newUri = new URI("file:" + uri)
+      if (updatedURLs.contains(newUri.toURL)) {
+        updatedURLs.remove(updatedURLs.indexOf(newUri.toURL))
+      }
+    })
+    lead.urlclassloader = new ExtendibleURLClassLoader(lead.urlclassloader.getParent)
```
Comment (kneeraj) [inline cluster/src/main/scala/io/snappydata/ToolsCallbackImpl.scala]: Please test the case where a secondary lead is brought in after a system is up and then lead becomes the primary leader. ( Not for this merge, but please note this )

File: cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala (hunk 2 above)
Comment (smahajan05) [reply cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala]: For deploy package also, this deletes the jar file copied to executor/servers work directory

File: cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala [hunk 4]
```diff
@@ -178,6 +176,30 @@ class SnappyExecutor(
...
+          logDebug(s"Deleted jarFile $jarFile")
+        }
+        var updatedURLs = urlClassLoader.getURLs().toBuffer
+        updatedURLs.foreach(url => {
+          if (url != null && url.toString.contains(jarFile.toString)) {
```
[the remark this answers is not in the document]
Comment (smahajan05) [reply cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala]: If the url file from urlclassloader is = file:/home/smahajan/snappydata/build-artifacts/scala-2.11/snappy/work/localhost-server-1/./com.datastax.spark_spark-cassandra-connector_2.11-2.0.7.jar and jarfile to be deleted from executor is = ./com.datastax.spark_spark-cassandra-connector_2.11-2.0.7.jar. so i am checking that the url file path contains jarfile path. If it is present then deleting the file. This is similar for deploy packages also.

No need of url != null. will remove that

File: core/src/main/scala/org/apache/spark/sql/execution/ddl.scala (hunk 1 above)
Comment (smahajan05) [reply core/src/main/scala/org/apache/spark/sql/execution/ddl.scala]: Will try to remove this duplicate code in later commits.

File: cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala [hunk 5]
```diff
@@ -178,6 +176,30 @@ class SnappyExecutor(
...
+            updatedURLs.remove(updatedURLs.indexOf(url))
+            val parent = Thread.currentThread().getContextClassLoader
+            updatedURLs.foreach(url => urlClassLoader.addURL(url))
```
[the remark this answers is not in the document]
Comment (smahajan05) [reply cluster/src/main/scala/org/apache/spark/executor/SnappyExecutor.scala]: For every jarFile i am searching in updatedURLS and after finding it getting the index for that jarFile from updatedURLS and then deleting the file.
No need of second updatedURLS.foreach loop. Removing it.

File: core/src/dunit/scala/io/snappydata/cluster/CassandraSnappyDUnitTest.scala [hunk 6]
```diff
@@ -191,6 +199,114 @@ class CassandraSnappyDUnitTest(val s: String)
...
+    count = 0
+    if (rs ne null) {
+      while (rs.next()) {
+        count += 1
+      }
+      rs.close()
+    }
+    assert(count == 0)
+
+    val jarPath = s"$snappyProductDir/jars/hadoop-client-2.7.7.jar"
```
Comment (ashetkar) [inline core/src/dunit/scala/io/snappydata/cluster/CassandraSnappyDUnitTest.scala]: Please test the deployment of the jar via another job or something, may be later if you have already manually tested it.

Comment (smahajan05) [reply core/src/dunit/scala/io/snappydata/cluster/CassandraSnappyDUnitTest.scala]: Ok will do in later commits

**Pull request merged**